crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -1226,6 +1255,80 @@ impl AuthorityPerEpochStore {
         }
     }
 
+    pub fn get_stored_execution_time_observations(
+        protocol_config: &ProtocolConfig,
+        committee: Arc<Committee>,
+        object_store: &dyn ObjectStore,
+    ) -> impl Iterator<Item = (AuthorityIndex, u64, ExecutionTimeObservationKey, Duration)> {
+        if protocol_config.per_object_congestion_control_mode()
+            != PerObjectCongestionControlMode::ExecutionTimeEstimate
+        {
+            return itertools::Either::Left(std::iter::empty());
+        }
+
+        // Load stored execution time observations from the SuiSystemState object.
+        let system_state =
+            sui_system_state::get_sui_system_state(object_store).expect("System state must exist");
+        let system_state = match system_state {
+            SuiSystemState::V2(system_state) => system_state,
+            SuiSystemState::V1(_) => {
+                error!("`PerObjectCongestionControlMode::ExecutionTimeEstimate` cannot load execution time observations to SuiSystemState because it has an old version. This should not happen outside tests.");
+                return itertools::Either::Left(std::iter::empty());
+            }
+            #[cfg(msim)]
+            SuiSystemState::SimTestV1(_)
+            | SuiSystemState::SimTestShallowV2(_)
+            | SuiSystemState::SimTestDeepV2(_) => {
+                return itertools::Either::Left(std::iter::empty());
+            }
+        };
+        let Ok::<Vec<u8>, _>(stored_observations_bytes) = get_dynamic_field_from_store(
+            object_store,
+            system_state.extra_fields.id.id.bytes,
+            &EXTRA_FIELD_EXECUTION_TIME_ESTIMATES_KEY,
+        ) else {
+            warn!("Could not find stored execution time observations. This should only happen in the first epcoh where ExecutionTimeEstimate mode is enabled.");
+            return itertools::Either::Left(std::iter::empty());
+        };
+        // TODO-DNS is the double-serialize necessary to store as a Vec<u8> in Move?
```

Comment:
DNS

Reply:
i'm not sure? I think probably yes? otherwise the contents of the field will not really be a Move type which seems bad